### ISSUE TITLE
[Feature]: retain trace for first failure only, not including retries

### ISSUE THREAD

**Issue by JacobCrofts** [microsoft/playwright#29531]
### 🚀 Feature Request

I would like to add a new option for [trace recording](https://playwright.dev/docs/trace-viewer#recording-a-trace-on-ci): `on-first-failure`. This would record and retain a trace only during the first failure of a given test, not during retries.

Related but different: https://github.com/microsoft/playwright/issues/13797

### Example

```js
import { defineConfig } from '@playwright/test';
export default defineConfig({
  retries: 1,
  use: {
    // if a test fails, it will retry once, but a trace will only be collected for the first failure
    trace: 'on-first-failure',
  },
});
```

### Motivation

My team uses Playwright for end-to-end tests that cause state modifications within our app. In theory, our tests "clean up" after themselves after each test completes, or fails, but implementing such functionality perfectly is difficult, especially for longer tests that fail in between two state changes.

We have run into issues like the following:

1. a test fails on the first try
2. the test tries to clean up after itself, but separately fails to do so
3. the test fails on the first _retry_ as a side effect of the failed cleanup step

The option I am suggesting would be preferable to:

- `on-first-retry` because it captures the original failure rather than a side effect of the original failure
- `retain-on-failure` because it saves space

I would be happy to implement this myself; I'll just wait a few days for feedback or a green light.

**Comment by JacobCrofts** [comment]
@yury-s Thanks for the tip. What you suggested would help us cut down on the volume of logs we collect, but I see that @StevanFreeborn has so kindly already implemented the feature I suggested, and that's my preferred approach. I think we're good to close this? (or perhaps it needs a release label first)

**Comment by dgozman** [comment]
This has been released in v1.43.
